core/src/main/scala/geotrellis/store/GeoTrellisRasterSourceLegacy.scala

```diff
@@ -86,48 +144,86 @@ class GeoTrellisRasterSourceLegacy(
     attributeStore.
       layerIds.
       filter(_.name == layerName).
       sortWith(_.zoom > _.zoom).
       toStream. // We will be lazy about fetching higher zoom levels
       map { id =>
-        val metadata = attributeStore.readMetadata[TileLayerMetadata[SpatialKey]](id)
-        Layer(id, metadata, bandCount)
+        val metadata = attributeStore.readMetadataErased(id)
+        LayerLegacy(id, metadata, bandCount)
       }
   }
 
-  def readTiles(reader: CollectionLayerReader[LayerId], layerId: LayerId, extent: Extent, bands: Seq[Int]): Seq[(SpatialKey, MultibandTile)] with Metadata[TileLayerMetadata[SpatialKey]] = {
+  def readTiles(
+    reader: CollectionLayerReader[LayerId],
+    layerId: LayerId,
+    extent: Extent,
+    bands: Seq[Int],
+    time: Option[ZonedDateTime] = None
+  ): Seq[(SpatialKey, MultibandTile)] with Metadata[TileLayerMetadata[SpatialKey]] = {
     def spatialTileRead =
       reader.query[SpatialKey, Tile, TileLayerMetadata[SpatialKey]](layerId)
         .where(Intersects(extent))
         .result
-        .withContext(tiles =>
-          // Convert single band tiles to multiband
-          tiles.map { case (key, tile) => (key, MultibandTile(tile)) }
-        )
+        .withContext { _.map { case (key, tile) => (key, MultibandTile(tile)) } }
 
     def spatialMultibandTileRead =
       reader.query[SpatialKey, MultibandTile, TileLayerMetadata[SpatialKey]](layerId)
         .where(Intersects(extent))
         .result
-        .withContext(tiles =>
-          tiles.map { case (key, tile) => (key, tile.subsetBands(bands)) }
-        )
+        .withContext { _.map { case (key, tile) => (key, tile.subsetBands(bands)) } }
+
+    def spaceTimeTileRead = {
+      val query =
+        reader
+          .query[SpaceTimeKey, Tile, TileLayerMetadata[SpaceTimeKey]](layerId)
+          .where(Intersects(extent))
+
+      time
+        .fold(query)(t => query.where(At(t)))
+        .result
+        .withContext { _.map { case (key, tile) => (key, MultibandTile(tile)) } }
+        .toSpatial
+    }
+
+    def spaceTimeMultibandTileRead = {
+      val query =
+        reader
+          .query[SpaceTimeKey, MultibandTile, TileLayerMetadata[SpaceTimeKey]](layerId)
+          .where(Intersects(extent))
+
+      time
+        .fold(query)(t => query.where(At(t)))
+        .result
+        .withContext { _.map { case (key, tile) => (key, tile.subsetBands(bands)) } }
+        .toSpatial
+    }
 
     val header = reader.attributeStore.readHeader[LayerHeader](layerId)
 
     if (!header.keyClass.contains("spark")) {
       (Class.forName(header.keyClass), Class.forName(header.valueClass)) match {
-        case (SpatialKeyClass, TileClass) => spatialTileRead
-        case (SpatialKeyClass, MultibandTileClass) => spatialMultibandTileRead
+        case (SpatialKeyClass, TileClass)            => spatialTileRead
+        case (SpatialKeyClass, MultibandTileClass)   => spatialMultibandTileRead
```

Comment:
I definitely feel bad about GeoTrellisRasterSource

Reply:
Because of this match or other reasons too ?